### ISSUE TITLE
history/save should be somewhat overhauled

### ISSUE THREAD

**Issue by GonziHere** [darktable-org/darktable#13256]
**Is your feature request related to a problem? Please describe.**
There are several issues compounding into bigger problem when it comes to the handling of the history. Programs typically let you drag something with mouse and "preview" the changes, but only letting go of said mouse will create standard history entry (for ctrl-z undo action I mean). Alternatively, hitting escape will cancel it fully.
DT doesn't do that. If you simply try some slider and decide that it doesn't work, you'll still have 10 steps in your history. It's annoying.
There is a system of history steps that's weird/hard to understand. It's not clear when and why it creates next entry, and I was surprised that I cannot click and edit some "history step" without loosing the future edits from that point. I kinda get it, but I'd like it to be more like a branch. I'd like to keep my edit history, just grayed out or something.
I'd also like to be able to save manually via some option (and prompt about unsaved progress). Right now, it's possible to click 10 steps into the history (to say, create snapshot), misclick elsewhere to create some edit (hell, just scroll the wheel on an input) and the history is gone AND saved.

**Describe the solution you'd like**
History (undo action) should contain only actual commited user actions (not all the moments of me dragging my mouse on slider).
History stack should work differently. Possibly show each actual commited action but surely with the ability to tweak, or at least remove an entry from the list. It should also be harder to loose the history.

**Alternatives**
The History stack could maybe be simply communicated differently or something, but as of now, it's misleading.

**Additional context**
I can technically use DSLRs and I heavily use one, but mainly for last 2 years, to capture my kid (as in, not a pro). I was fine with Nikon NX Studio until I wasn't, and I've moved to DT. I love it. It has a steep learning curve, but I love it. I've spend significant time on this Christmas to learn it more and I dig it. That being said, I still have the perspective of someone who isn't used to all it's quirks. I can see the philosophy for many things and again, generally, I like it. But the history stack, action history and active modules (without visible order of them) is very hard to grasp. It took a few hours of trial end error to get a feel for how it works and now, that I have it, I still don't think it's well designed. It's quirks make unintentional errors easy, therefore it's hard to give an access to less tech savvy people in family.

**Comment by TurboGit** [comment]
> If you simply try some slider and decide that it doesn't work, you'll still have 10 steps in your history. It's annoying.

How? Why it is annoying?

> There is a system of history steps that's weird/hard to understand. 

That's a common way to handle history. See Lr for example. And be sure to have a look at the documentation.

> I kinda get it, but I'd like it to be more like a branch.

I can assure you that there is probably zero chance for this to happen. To have multiple edits for the same picture one has to use duplicate.

> History stack should work differently. Possibly show each actual commited action but surely with the ability to tweak, or at least remove an entry from the list.

Just compress history.

As a note for possible future enhancement, I'm contemplating the possibility to restore the history of a snapshot.

**Comment by GonziHere** [comment]
> >  If you simply try some slider and decide that it doesn't work, you'll still have 10 steps in your history. It's annoying.

> How? Why it is annoying?
I would expect to open up an image, take a white balance slider and while holding the mouse down, I'll move it back and forth for a minute. In the end, I'll either let go of the mouse (and 'commit' that change) or I'll hit ESC to cancel it altogether, deciding that the captured WB was good. 

In DT, this minute of me dragging slider back and forth will spam-create 60 undo history entries. It's the only program I've ever encountered that works like this. Maybe Adobe also does that, but most software works as I've described (including several image editors, game engines, and Blender).

> I'm contemplating the possibility to restore the history of a snapshot.

That's technically there with duplicates, no?

My issue was that I've randomly clicked to some history step just to preview it, then dragged some settings (I did it - wrongly - to play with white balance without applied tone mapping) and I've lost said tone mapping. It was easy to fix, but It was unexpected from the user perspective and now that I know how it works, I still somewhat don't like it. 

There should be some separation between preview of the history (which is a great feature) and undoing a few steps to try a different edit. Right now, that distinction isn't obvious IMO. Especially since history is on the left, with the less dangerous options, not with "editors" on the right.

So, maybe disable edits without explicit "go back to this history version", maybe create said snapshot at the moment of going back in history, maybe something else, but right now, it's suboptimal from my POV.

PS: All of those issues are significantly amplified by the instant-autosave feature.

**Comment by jenshannoschwalm** [comment]
> In DT, this minute of me dragging slider back and forth will spam-create 60 undo history entries.

This should not happen. If you keep changing a single value/slider in a module the history will take the new setting but there are no new entries in the history. This could happen many dt versions ago ...

Iy you do changes on another parameter/slider - yes a new history entry.

Are you really sure about this?

**Comment by GonziHere** [comment]
I've meant that as the "undo buffer" part for ctrl-z (also hitting esc while dragging doesn't cancel mouse drag and doesn't restore the original state). Both are standard/expected behaviors in basically everything I've ever used, which is why I'm surprised that it isn't there.

History stack works as you've described :-) .

**Comment by eoyilmaz** [comment]
TL;DR: Ctrl + Z should go one step back in the History...

I agree with @GonziHere on the Undo (Ctrl +Z) behaviour. The undo queue (not the history ) is not working as expected, I expect hitting Ctrl + Z / Cmd +Z to go back the to state that I first pressed the mouse button, completely skipping all the in between states created while I was dragging my mouse.

Here is a video, you will hear when I pressed and released the mouse button:

https://user-images.githubusercontent.com/1786804/213584914-6ad7ff3d-61de-48ff-9215-0ebc830c85be.mp4

1. I press the mouse button
2. drag the control between several position
3. and then release the mouse button
4. and then I had to hit Ctrl + Z several times just to go to the state I first pressed the button

Expected behaviour is; a single Ctrl + Z press goes to the state that the mouse button first pressed.

**Comment by GonziHere** [comment]
While I think that I provide valid feedback for the whole autosave/history stack system from POV of new user, I can appreciate that it's too complex of a change.
Ultimately, I'm not a 'new user' anymore, so **it's not a hill I'd die on**, if others don't agree.

Nevertheless, the Ctrl - Z issue bugs me constantly and I do consider it a bug.

Ctrl-Z is undo of an **action** as per convention. While **holding the mouse, I'm only previewing** said action and ESC cancels it altogether. The only write to the undo buffer should happen when I'll let go of the mouse - that's the moment where I'm actually committing the action.

Photoshop does it wrong and since DT isn't it's replacement, IDK why this bug should be copied.

Thanks for consideration.

**Comment by eoyilmaz** [comment]
I still think Ctrl+Z (Cmd+Z) should restore the state between mouse clicks... not inbetween, now that I can build Darktable on macos, I'll look in to a proposed implementation... soon... hopefully...

**Comment by ralfbrown** [comment]
The frequency of undo records was reduced in 4.4, which at least partially addresses this old request.
